std/algorithm/comparison.d

```diff
@@ -732,78 +731,133 @@ template equal(alias pred = "a == b")
     See_Also:
         $(HTTP sgi.com/tech/stl/_equal.html, STL's _equal)
     +/
-    bool equal(Range1, Range2)(Range1 r1, Range2 r2)
-    if (isInputRange!Range1 && isInputRange!Range2 &&
-        is(typeof(binaryFun!pred(r1.front, r2.front))))
+    bool equal(R, Ss...)(R r, Ss ss)
+    if (ss.length >= 1 &&
+        allSatisfy!(isInputRange, R, Ss) &&
+        (!isInfinite!R || !allSatisfy!(isInfinite, Ss)) // at least one must be finite
+        )
     {
-        static assert(!(isInfinite!Range1 && isInfinite!Range2),
-            "Both ranges are known to be infinite");
-
-        //No pred calls necessary
-        static if (isEmptyRange!Range1 || isEmptyRange!Range2)
-        {
-            return r1.empty && r2.empty;
-        }
-        else static if ((isInfinite!Range1 && hasFixedLength!Range2) ||
-            (hasFixedLength!Range1 && isInfinite!Range2))
-        {
-            return false;
-        }
-        //Detect default pred and compatible dynamic array
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isArray!Range1 && isArray!Range2 && is(typeof(r1 == r2)))
-        {
-            return r1 == r2;
-        }
-        // if one of the arguments is a string and the other isn't, then auto-decoding
-        // can be avoided if they have the same ElementEncodingType
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isAutodecodableString!Range1 != isAutodecodableString!Range2 &&
-            is(ElementEncodingType!Range1 == ElementEncodingType!Range2))
+        enum isEquableToR(T) = is(typeof({ return R.init == T.init; }));
+
+        // start by detecting default pred and compatible dynamic array
+        static if (is(typeof(pred) == string) &&
```

Comment:
I don't see a necessity for special casing this. Maybe only special case for two arrays exactly.